### PR TITLE
TDL-15274: Fix pagination for Stories and Tasks

### PR DESCRIPTION
**Pull request by hpatel41**
# Description of change
TDL-15274: Fix pagination for Stories and Tasks
- Added code change to cache all project ids for 'Stories' and 'Tasks' stream.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by dbshah1212 — CHANGES_REQUESTED** [review]
Project is also used for section stream, please add this same logic for section as well: https://github.com/singer-io/tap-asana/blob/master/tap_asana/streams/sections.py

**Comment by hpatel41** [comment]
> Project is also used for section stream, please add this same logic for section as well: https://github.com/singer-io/tap-asana/blob/master/tap_asana/streams/sections.py

Added caching of Project IDs for the section stream.

File: tap_asana/streams/stories.py [hunk 1]
```diff
@@ -59,14 +59,22 @@ def get_objects(self):
...
-        for task in self.call_api("tasks", project=project["gid"]): 
-          task_gid = task.get('gid')
-          for story in Context.asana.client.stories.get_stories_for_task(task_gid=task_gid, opt_fields=opt_fields):
-            session_bookmark = self.get_updated_session_bookmark(session_bookmark, story[self.replication_key])
-            if self.is_bookmark_old(story[self.replication_key]):
-              yield story
+        project_ids.append(project["gid"])
+
+    # iterate over all project ids and continue fetching
+    for project_id in project_ids:
```
Comment (dbshah1212) [inline tap_asana/streams/stories.py:70]: Add unit tests for this.

File: tap_asana/streams/tasks.py [hunk 2]
```diff
@@ -53,12 +53,20 @@ def get_objects(self):
...
+    project_ids = []
     for workspace in self.call_api("workspaces"):
       for project in self.call_api("projects", workspace=workspace["gid"]):
-        for task in self.call_api("tasks", project=project["gid"], opt_fields=opt_fields, modified_since=modified_since): 
-          session_bookmark = self.get_updated_session_bookmark(session_bookmark, task[self.replication_key])
-          if self.is_bookmark_old(task[self.replication_key]):
-            yield task
+        project_ids.append(project["gid"])
+
+    # iterate over all project ids and continue fetching
```
Comment (dbshah1212) [inline tap_asana/streams/tasks.py:62]: Add unit test for this.

File: tap_asana/streams/sections.py [hunk 3]
```diff
@@ -23,10 +23,17 @@ def get_objects(self):
...
+    # collect list of project ids
+    project_ids = []
     for workspace in self.call_api("workspaces"):
       for project in self.call_api("projects", workspace=workspace["gid"]):
-        for section in Context.asana.client.sections.get_sections_for_project(project_gid=project["gid"], owner="me", opt_fields=opt_fields):
-          yield section
+        project_ids.append(project["gid"])
+
+    # iterate on all project ids and execute rest of the sync
+    for project_id in project_ids:
```
Comment (dbshah1212) [inline tap_asana/streams/sections.py:30]: Add unit test for this.

File: tap_asana/streams/stories.py (hunk 1 above)
Comment (hpatel41) [reply tap_asana/streams/stories.py:70]: Added unittest

File: tap_asana/streams/tasks.py (hunk 2 above)
Comment (hpatel41) [reply tap_asana/streams/tasks.py:62]: Added unittest

File: tap_asana/streams/sections.py (hunk 3 above)
Comment (hpatel41) [reply tap_asana/streams/sections.py:30]: Added unittest

**Review by karanpanchal-crest — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tap_asana/streams/sections.py [hunk 4]
```diff
@@ -23,10 +23,17 @@ def get_objects(self):
...
     modified_since = bookmark.strftime("%Y-%m-%dT%H:%M:%S.%f")
     opt_fields = ",".join(self.fields)
+
+    # collect list of project ids
+    project_ids = []
     for workspace in self.call_api("workspaces"):
       for project in self.call_api("projects", workspace=workspace["gid"]):
-        for section in Context.asana.client.sections.get_sections_for_project(project_gid=project["gid"], owner="me", opt_fields=opt_fields):
-          yield section
+        project_ids.append(project["gid"])
```
Comment (KrisPersonal) [inline tap_asana/streams/sections.py:27]: How big this list will be? If is a big list this can potentially be an issue as you keep appending.

Comment (hpatel41) [reply tap_asana/streams/sections.py:27]: It depends from client to client. But, for the safe side, we have added the yielding of project ids.

**Review by kspeer825 — COMMENTED** [review]
Can this issue be reproduced via tap-tester test?

**Review by KrisPersonal — APPROVED** [review]

**Comment by hpatel41** [comment]
> Can this issue be reproduced via tap-tester test?

@kspeer825 We cannot reproduce this issue in the tap-tester, as in the first loop we get the projects on which we need to fetch the required data, based on the project ids.

To fetch the data for the next page we require an offset token which is valid for 15-20 minutes. Hence, the inner loops should have enough data to keep iterating over the project ids before fetching the next page for projects.

Reference: https://developers.asana.com/docs/pagination

File: .circleci/config.yml [hunk 5]
```diff
@@ -12,6 +12,17 @@ jobs:
...
+          command: |
+            source /usr/local/share/virtualenvs/tap-asana/bin/activate
+            pip install nose coverage
+            nosetests --with-coverage --cover-erase --cover-package=tap_asana --cover-html-dir=htmlcov tests/unittests
+            coverage html
+      - store_test_results:
+          path: test_output/report.xml
+      - store_artifacts:
+          path: htmlcov
       - add_ssh_keys
```
Comment (kspeer825) [inline .circleci/config.yml:32]: The `add_ssh_keys` is unnecessary. Also please update to the latest tap-tester image and use the tap_tester_sandbox environment variables over the sandbox ones.

**Review by kspeer825 — CHANGES_REQUESTED** [review]
> Please merge the regression suite PR #39 to master which is approved. And then  merge master back into this PR. I would like to understand why we do not see the standard pagination test fail for this issue. We should be able to write a failing tap-tester test for this case and then see it pass with the code fix here. I do not think this should be merged until that is done.

File: .circleci/config.yml (hunk 5 above)
Comment (hpatel41) [reply .circleci/config.yml:32]: Removed `add_ssh_keys`. In PR #39, we have updated the tap tester image, sandbox, and the command to run the integration tests.

**Comment by hpatel41** [comment]
> Please merge the regression suite PR #39 to master which is approved. And then merge master back into this PR. I would like to understand why we do not see the standard pagination test fail for this issue. We should be able to write a failing tap-tester test for this case and then see it pass with the code fix here. I do not think this should be merged until that is done.

@kspeer825 We require more data for failing the tests, but for 3000 records, the tap takes 1.5 minutes. Hence, to make the loop iterate for around 15 minutes, we will need to generate a huge amount of data. So, we are working to generate more data to fail the test cases.

**Review by kspeer825 — APPROVED** [review]
This looks great, thanks for updating.

**Pull request merged**